### ISSUE TITLE
Spec: Deep ModelEvaluator Module

### ISSUE THREAD

**Issue by Igoreduardobraga** [Igoreduardobraga/BreakEvenNLP#1]
# Spec: Deep ModelEvaluator Module

**Triage Label:** `ready-for-agent`  
**Status:** Ready for Implementation  
**Area:** Model Inference Pipeline  

---

## Problem Statement

Researchers running determinism and stability investigations in NLP have to maintain six separate model runner functions in `main.py` (`run_flan_t5`, `run_llama2`, `run_llama3`, `run_mistral`, `run_zephyr`, `run_chatgpt`). Each function duplicates batch loops, GPU tensor transfers, token generation, generated slice decoding, and label parsing regexes. Adding or updating models requires copying dozens of lines of procedural boilerplate, return signatures are inconsistent across models (some return 2 elements, others return 3), and testing model evaluation currently requires running the full 1200-line pipeline with heavy GPU weights.

## Solution

Consolidate model inference under a single deep `ModelEvaluator` module in `BreakEvenNLP/evaluator.py`. The module presents a single, uniform interface (`.evaluate(dataset, mode)`) returning `(golden, predicted, decodeds)`. Model family variations (Seq2Seq, CausalLM with chat templates, and API calls) and label parsing are encapsulated behind internal adapters, allowing `main.py` and instruction-tuning evaluation to evaluate any supported model through one predictable seam.

## User Stories

1. As a researcher investigating randomness factors, I want to evaluate any supported model via a single uniform method call, so that I don't need model-specific branching in my experiment loops.
2. As a researcher, I want `ModelEvaluator` to accept pre-loaded model and tokenizer instances, so that heavy model weights remain in GPU VRAM across all cross-validation folds without being reloaded from disk.
3. As a developer writing tests, I want `ModelEvaluator` to always return a standardized 3-tuple `(golden, predicted, decodeds)`, so that caller code never has to inspect tuple length or handle missing outputs conditionally.
4. As a researcher adding a new causal language model, I want chat template application and token slicing to be handled internally by a dedicated adapter, so that I don't have to duplicate batch loops and token decoding logic.
5. As a researcher running instruction tuning experiments, I want to evaluate the tuned model using the same `ModelEvaluator`, so that evaluation metrics and parsing behavior are 100% consistent across pre-trained and fine-tuned models.
6. As a developer running automated tests, I want to test the full evaluation and parsing pipeline using lightweight dummy models, so that I can verify evaluation correctness in milliseconds without a GPU.
7. As a researcher running prompting experiments, I want to specify `mode='prompting'` without in-context examples, so that the model is evaluated purely on zero-shot instructions.
8. As a researcher running in-context learning experiments, I want to specify `mode='icl'` with context shots, so that few-shot demonstrations are automatically formatted and injected into the prompt.
9. As a researcher testing Seq2Seq models like Flan-T5, I want the evaluator to handle encoder-decoder generation and `batch_decode` internally, so that I don't have to manage Seq2Seq-specific generation calls in `main.py`.
10. As a researcher testing CausalLM models like LLaMA-3, Mistral, and Zephyr, I want the evaluator to automatically slice off the prompt tokens and decode only newly generated tokens, so that prompt text does not corrupt label parsing.
11. As a researcher using the ChatGPT API, I want the evaluator to handle rate-limiting retries and persist partial results to disk, so that network errors or quota limits don't lose experimental progress.
12. As a maintainer, I want label parsing regexes to live exclusively inside the evaluation module, so that changes to class label extraction are localized to one place.
13. As a maintainer, I want `ModelEvaluator` to accept batch size, prompt format, and max new tokens as explicit parameters, so that evaluation does not depend on global CLI variables.
14. As a researcher reviewing experiment logs, I want any parsing failures to be consistently flagged with index `-1`, so that parsing failure rates can be audited across models and prompt formats.

## Implementation Decisions

- The evaluation logic is extracted into a dedicated deep module located at `BreakEvenNLP/evaluator.py`.
- The public interface of the module is:
  ```python
  ModelEvaluator(
      model,
      tokenizer,
      model_name: str,
      batch_size: int = 64,
      prompt_format: int = 0,
      max_new_tokens: int = 10,
      generation_config = None,
      device: str = "cuda"
  )
  ```
- The execution entry point is `evaluate(dataset, mode: str)` where `mode` is `'prompting'` or `'icl'`, returning `tuple[list[int], list[int], list[str]]` corresponding to `(golden, predicted, decodeds)`.
- Behind the seam, the module delegates to internal family adapters:
  - `_Seq2SeqAdapter` for encoder-decoder models (`flan-t5`)
  - `_CausalLMAdapter` for autoregressive chat models (`llama2`, `llama3`, `mistral`, `zephyr`)
  - `_APIAdapter` for external API models (`chatgpt`)
- Label parsing is encapsulated as a private helper `_parse_results(text, classes)` within `evaluator.py`, returning `-1` if no single class is matched.
- The existing runner functions in `main.py` (`run_flan_t5`, `run_llama2`, `run_llama2_optimized`, `run_llama3`, `run_mistral`, `run_zephyr`, `run_chatgpt`) and prompt helpers (`prepare_flan_t5_*`, `prepare_llama2_*`, `prepare_chatgpt_*`) are deleted and replaced by `ModelEvaluator`.
- `instruction_tuning_experiment` in `main.py` is updated to instantiate `ModelEvaluator` on the tuned model to evaluate both `prompting` and `icl` modes.

## Testing Decisions

- A good test verifies only external observable behavior across the seam: passing a dataset and model, and verifying that the returned `golden`, `predicted`, and `decodeds` lists have the correct values, lengths, and types.
- Tests do not assert on private adapter dispatch, internal regex calls, or private method execution.
- The test suite will be introduced in `BreakEvenNLP/tests/test_evaluator.py`.
- Prior art: the codebase currently has 0 automated tests; this test suite establishes the primary testing seam for model evaluation using simulated tokenizers and models (e.g. lightweight objects mimicking Hugging Face `generate` and `batch_decode`).

## Out of Scope

- Refactoring the dataset loading hierarchy (`TextDataset`, `ICLDataset`, `FineTuningDataset` in `data.py`) &mdash; addressed by Candidate 3 in a future iteration.
- Refactoring prompt construction into `PromptFormatter` &mdash; addressed by Candidate 2 in a future iteration.
- Encapsulating random number generators (`RNGController`) &mdash; addressed by Candidate 4 in a future iteration.
- Modifying fine-tuning training loops (`ft_experiment`, `SFTTrainer`).

## Further Notes

- Preserves exact model outputs, generation arguments (`do_sample=False`, `max_new_tokens=10`), and F1 scores.
- When `gh` CLI is installed and authenticated, this spec can be published directly to GitHub issues under the `ready-for-agent` label.

**Comment by Igoreduardobraga** [comment]
All 4 implementation tickets (#2, #3, #4, #5) have been completed and verified. ModelEvaluator module is fully functional, integrated into main.py, and covered by automated unit tests.
